clinical/src/main/java/com/hartwig/actin/clinical/curation/CurationDatabaseValidator.java

```diff
@@ -0,0 +1,119 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hartwig.actin.clinical.curation.config.IntoleranceConfig;
+import com.hartwig.actin.clinical.curation.config.NonOncologicalHistoryConfig;
+import com.hartwig.actin.clinical.curation.config.PrimaryTumorConfig;
+import com.hartwig.actin.clinical.curation.config.SecondPrimaryConfig;
+import com.hartwig.actin.clinical.datamodel.PriorOtherCondition;
+import com.hartwig.actin.doid.DoidModel;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+class CurationDatabaseValidator {
+
+    private static final Logger LOGGER = LogManager.getLogger(CurationDatabaseValidator.class);
+
+    static final String GENERIC_PARENT_DOID = "4"; // "disease"
+    static final String CANCER_PARENT_DOID = "14566"; // "disease of cellular proliferation"
+
+    @NotNull
+    private final DoidModel doidModel;
+
+    public CurationDatabaseValidator(@NotNull final DoidModel doidModel) {
+        this.doidModel = doidModel;
+    }
+
+    public void validate(@NotNull CurationDatabase database) {
+        LOGGER.info("Running curation database validation");
+
+        boolean allPrimaryTumorConfigsValid = validatePrimaryTumorConfigs(database.primaryTumorConfigs(), doidModel);
+        boolean allSecondPrimaryConfigsValid = validateSecondPrimaryConfigs(database.secondPrimaryConfigs(), doidModel);
+        boolean allNonOncologicalHistoryConfigsValid =
+                validateNonOncologicalHistoryConfigs(database.nonOncologicalHistoryConfigs(), doidModel);
+        boolean allIntoleranceConfigsValid = validateIntoleranceConfigs(database.intoleranceConfigs(), doidModel);
+
+        if (allPrimaryTumorConfigsValid && allSecondPrimaryConfigsValid && allNonOncologicalHistoryConfigsValid
+                && allIntoleranceConfigsValid) {
+            LOGGER.info(" Curation database validation found no quality issues");
+        }
+    }
+
+    @VisibleForTesting
+    static boolean validatePrimaryTumorConfigs(@NotNull List<PrimaryTumorConfig> primaryTumorConfigs,
+            @NotNull DoidModel doidModel) {
+        boolean allValid = true;
+        for (PrimaryTumorConfig primaryTumorConfig : primaryTumorConfigs) {
+            if (!hasValidDoids(primaryTumorConfig.doids(), doidModel, CANCER_PARENT_DOID)) {
+                allValid = false;
+                LOGGER.warn(" Invalid primary tumor doids configured for '{}': {}", primaryTumorConfig.input(), primaryTumorConfig.doids());
+            }
+        }
+        return allValid;
+    }
+
+    @VisibleForTesting
+    static boolean validateSecondPrimaryConfigs(@NotNull List<SecondPrimaryConfig> secondPrimaryConfigs,
+            @NotNull DoidModel doidModel) {
+        boolean allValid = true;
+        for (SecondPrimaryConfig secondPrimaryConfig : secondPrimaryConfigs) {
+            if (!hasValidDoids(secondPrimaryConfig.curated().doids(), doidModel, CANCER_PARENT_DOID)) {
```

Comment:
`SecondPrimaryConfig::curated` is nullable, so we should use an Optional here. We could use the same pattern I use below for the history without the filter and cast steps.